**Stack Overflow Developer Survey 2025 Analysis**
**Опис проєкту**

Метою цього проєкту є аналіз великого датасету Stack Overflow Developer Survey 2025 для дослідження тенденцій сучасного IT-ринку, популярності технологій, рівня компенсацій, форматів роботи та професійного розвитку спеціалістів.

У межах аналізу було досліджено відповіді понад 49 тисяч респондентів із різних країн світу. Проєкт демонструє практичну роботу з великим датасетом: завантаження даних, перевірку якості, аналіз пропущених значень, фільтрацію, групування, агрегацію та формування аналітичних висновків.

**Основні питання аналізу**
- Який обсяг вибірки представлений у датасеті?
- Наскільки повними є відповіді респондентів?
- Який типовий досвід роботи мають респонденти?
- Наскільки поширений remote-формат роботи?
- Яку частку респондентів становлять спеціалісти, що використовують Python?
- Як відрізняється компенсація Python-розробників залежно від країни?
- Який звʼязок можна спостерігати між освітою та компенсацією?
- У яких індустріях найчастіше працюють високооплачувані remote-спеціалісти?
  
**Інструменти**
- Python
- Pandas
- Jupyter Notebook
- Exploratory Data Analysis
- Data Cleaning
- Data Aggregation
- Data Interpretation

# 0. Підготовка середи за допомогою імпорту бібліотеки Pandas

**0.1. Імпорт бібліотеки Pandas та додаткові налаштування**

In [1]:
import pandas as pd

In [2]:
# Відображення всіх стовпців DataFrame, щоб бачити повну структуру даних
pd.set_option('display.max_columns', None)

# Додатково збільшимо ширину відображення
pd.set_option('display.width', 1000)

# Та відобразимо повний текст в комірках
pd.set_option('display.max_colwidth', None)

**0.2. Завантаження вихідних даних та перевірка завантаження**

Завантаження попередньо скачаних файлів формату csv з офіційного архіву Stack Overflow Developer Survey 2025 та розміщених локально:
- survey_results_public.csv: основний датасет з відповідями респондентів;
- survey_results_schema.csv: схема, опис питань та технічних назв колонок

In [3]:
df = pd.read_csv('survey_results_public.csv', encoding='utf-8')
df1 = pd.read_csv('survey_results_schema.csv', encoding='utf-8')

Перевіряємо коректність завантаження

In [4]:
df.shape

(49123, 170)

In [5]:
df1.shape

(139, 6)

**0.3. Первинне дослідження структури датасетів**

In [6]:
# Дивимось перші 5 рядків основного датасету
df.head()

,ResponseId,MainBranch,Age,EdLevel,Employment,EmploymentAddl,WorkExp,LearnCodeChoose,LearnCode,LearnCodeAI,AILearnHow,YearsCode,DevType,OrgSize,ICorPM,RemoteWork,PurchaseInfluence,TechEndorseIntro,TechEndorse_1,TechEndorse_2,TechEndorse_3,TechEndorse_4,TechEndorse_5,TechEndorse_6,TechEndorse_7,TechEndorse_8,TechEndorse_9,TechEndorse_13,TechEndorse_13_TEXT,TechOppose_1,TechOppose_2,TechOppose_3,TechOppose_5,TechOppose_7,TechOppose_9,TechOppose_11,TechOppose_13,TechOppose_16,TechOppose_15,TechOppose_15_TEXT,Industry,JobSatPoints_1,JobSatPoints_4,JobSatPoints_5,JobSatPoints_6,JobSatPoints_7,JobSatPoints_8,JobSatPoints_9,JobSatPoints_10,JobSatPoints_11,JobSatPoints_13,JobSatPoints_14,JobSatPoints_15,JobSatPoints_16,JobSatPoints_15_TEXT,AIThreat,NewRole,ToolCountWork,ToolCountPersonal,Country,Currency,CompTotal,LanguageChoice,LanguageHaveWorkedWith,LanguageWantToWorkWith,LanguageAdmired,LanguagesHaveEntry,LanguagesWantEntry,DatabaseChoice,DatabaseHaveWorkedWith,DatabaseWantToWorkWith,DatabaseAdmired,DatabaseHaveEntry,DatabaseWantEntry,PlatformChoice,PlatformHaveWorkedWith,PlatformWantToWorkWith,PlatformAdmired,PlatformHaveEntry,PlatformWantEntry,WebframeChoice,WebframeHaveWorkedWith,WebframeWantToWorkWith,WebframeAdmired,WebframeHaveEntry,WebframeWantEntry,DevEnvsChoice,DevEnvsHaveWorkedWith,DevEnvsWantToWorkWith,DevEnvsAdmired,DevEnvHaveEntry,DevEnvWantEntry,SOTagsHaveWorkedWith,SOTagsWantToWorkWith,SOTagsAdmired,SOTagsHaveEntry,SOTagsWant Entry,OpSysPersonal use,OpSysProfessional use,OfficeStackAsyncHaveWorkedWith,OfficeStackAsyncWantToWorkWith,OfficeStackAsyncAdmired,OfficeStackHaveEntry,OfficeStackWantEntry,CommPlatformHaveWorkedWith,CommPlatformWantToWorkWith,CommPlatformAdmired,CommPlatformHaveEntr,CommPlatformWantEntr,AIModelsChoice,AIModelsHaveWorkedWith,AIModelsWantToWorkWith,AIModelsAdmired,AIModelsHaveEntry,AIModelsWantEntry,SOAccount,SOVisitFreq,SODuration,SOPartFreq,SO_Dev_Content,SO_Actions_1,SO_Actions_16,SO_Actions_3,SO_Actions_4,SO_Actions_5,SO_Actions_6,SO_Actions_9,SO_Actions_7,SO_Actions_10,SO_Actions_15,SO_Actions_15_TEXT,SOComm,SOFriction,AISelect,AISent,AIAcc,AIComplex,AIToolCurrently partially AI,AIToolDon't plan to use AI for this task,AIToolPlan to partially use AI,AIToolPlan to mostly use AI,AIToolCurrently mostly AI,AIFrustration,AIExplain,AIAgents,AIAgentChange,AIAgent_Uses,AgentUsesGeneral,AIAgentImpactSomewhat agree,AIAgentImpactNeutral,AIAgentImpactSomewhat disagree,AIAgentImpactStrongly agree,AIAgentImpactStrongly disagree,AIAgentChallengesNeutral,AIAgentChallengesSomewhat disagree,AIAgentChallengesStrongly agree,AIAgentChallengesSomewhat agree,AIAgentChallengesStrongly disagree,AIAgentKnowledge,AIAgentKnowWrite,AIAgentOrchestration,AIAgentOrchWrite,AIAgentObserveSecure,AIAgentObsWrite,AIAgentExternal,AIAgentExtWrite,AIHuman,AIOpen,ConvertedCompYearly,JobSat
0,1,I am a developer by profession,25-34 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Employed,"Caring for dependents (children, elderly, etc.)",8.0,"Yes, I am not new to coding but am learning new coding techniques or programming language","Online Courses or Certification (includes all media types);Other online resources (e.g. standard search, forum, online community)","Yes, I learned how to use AI-enabled tools for my personal curiosity and/or hobbies",AI CodeGen tools or AI-enabled apps,14.0,"Developer, mobile",20 to 99 employees,People manager,Remote,"Yes, I influenced the purchase of a substantial addition to the tech stack",Work,10.0,7.0,9.0,6.0,3.0,11.0,12.0,1.0,8.0,14.0,NaN,15.0,7.0,8.0,12.0,11.0,1.0,6.0,13.0,3.0,16.0,NaN,Fintech,3.0,9.0,5.0,10.0,12.0,11.0,2.0,6.0,7.0,13.0,14.0,15.0,8.0,NaN,I'm not sure,I have neither consider or transitioned into a new career or industry,7.0,3.0,Ukraine,EUR European Euro,52800.0,Yes,Bash/Shell (all shells);Dart;SQL,Dart,Dart,NaN,NaN,Yes,Cloud Firestore;PostgreSQL,NaN,NaN,NaN,NaN,Yes,Amazon Web Services (AWS);Cloudflare;Firebase;Google Cloud;Gradle,NaN,NaN,NaN,NaN,No,NaN,NaN,NaN,NaN,NaN,Ye

In [7]:
# Дивимось перші 5 рядків схеми питань
df1.head()

,qid,qname,question,type,sub,sq_id
0,QID18,TechEndorse_1,What attracts you to a technology or causes you to endorse it (most to least important)?,RO,AI integration or AI Agent capabilities,1.0
1,QID18,TechEndorse_2,What attracts you to a technology or causes you to endorse it (most to least important)?,RO,Easy-to-use API,2.0
2,QID18,TechEndorse_3,What attracts you to a technology or causes you to endorse it (most to least important)?,RO,Robust and complete API,3.0
3,QID18,TechEndorse_4,What attracts you to a technology or causes you to endorse it (most to least important)?,RO,Customizable and manageable codebase,4.0
4,QID18,TechEndorse_5,What attracts you to a technology or causes you to endorse it (most to least important)?,RO,Reputation for quality,5.0


In [8]:
# Перевіряємо назви колонок в основному датасеті
df.columns.tolist()

['ResponseId',
 'MainBranch',
 'Age',
 'EdLevel',
 'Employment',
 'EmploymentAddl',
 'WorkExp',
 'LearnCodeChoose',
 'LearnCode',
 'LearnCodeAI',
 'AILearnHow',
 'YearsCode',
 'DevType',
 'OrgSize',
 'ICorPM',
 'RemoteWork',
 'PurchaseInfluence',
 'TechEndorseIntro',
 'TechEndorse_1',
 'TechEndorse_2',
 'TechEndorse_3',
 'TechEndorse_4',
 'TechEndorse_5',
 'TechEndorse_6',
 'TechEndorse_7',
 'TechEndorse_8',
 'TechEndorse_9',
 'TechEndorse_13',
 'TechEndorse_13_TEXT',
 'TechOppose_1',
 'TechOppose_2',
 'TechOppose_3',
 'TechOppose_5',
 'TechOppose_7',
 'TechOppose_9',
 'TechOppose_11',
 'TechOppose_13',
 'TechOppose_16',
 'TechOppose_15',
 'TechOppose_15_TEXT',
 'Industry',
 'JobSatPoints_1',
 'JobSatPoints_4',
 'JobSatPoints_5',
 'JobSatPoints_6',
 'JobSatPoints_7',
 'JobSatPoints_8',
 'JobSatPoints_9',
 'JobSatPoints_10',
 'JobSatPoints_11',
 'JobSatPoints_13',
 'JobSatPoints_14',
 'JobSatPoints_15',
 'JobSatPoints_16',
 'JobSatPoints_15_TEXT',
 'AIThreat',
 'NewRole',
 'ToolCountWor

In [9]:
# Перевіряємо назви колонок в схемі питань
df1.columns.tolist()

['qid', 'qname', 'question', 'type', 'sub', 'sq_id']

In [10]:
# Перевіряємо "якість" основного датасету на предмет визначення типів даних, кількості непустих значень та загального розміру таблиці
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49123 entries, 0 to 49122
Columns: 170 entries, ResponseId to JobSat
dtypes: float64(50), int64(1), object(119)
memory usage: 63.7+ MB


In [11]:
# Розраховуємо кількість пропусків в кожній колонці
# isna() повертає True там, де значення пропущено;
# sum() рахує кількість True в кожній колонці.
# Виведемо перші 20 колонок з найбільшим числом пропусків.

missing_by_column = df.isna().sum().sort_values(ascending=False)
missing_by_column.head(20)

AIAgentObsWrite         48859
SOTagsWant Entry        48693
SOTagsHaveEntry         48666
AIModelsWantEntry       48649
AIAgentOrchWrite        48646
JobSatPoints_15_TEXT    48459
AIAgentKnowWrite        48358
AIModelsHaveEntry       48348
SO_Actions_15_TEXT      48300
AIAgentExtWrite         48265
CommPlatformWantEntr    47940
CommPlatformHaveEntr    47648
DatabaseWantEntry       47595
OfficeStackWantEntry    47506
TechOppose_15_TEXT      47478
TechEndorse_13_TEXT     47115
DevEnvWantEntry         47013
DatabaseHaveEntry       46975
OfficeStackHaveEntry    46536
WebframeWantEntry       46497
dtype: int64

In [12]:
# Сформуємо таблицю даних за колонками, щоб наочно побачити якість заповнення полів

data_quality = pd.DataFrame({
    'column': df.columns,
    'non_missing_values': df.notna().sum().values,
    'missing_values': df.isna().sum().values,
    'missing_percent': (df.isna().mean().values * 100)
})

# Відсортуємо пропуски за відсотком від більшого до меншого

data_quality = data_quality.sort_values('missing_percent', ascending=False)

data_quality.head(20)

,column,non_missing_values,missing_values,missing_percent
163,AIAgentObsWrite,264,48859,99.462574
96,SOTagsWant Entry,430,48693,99.124646
95,SOTagsHaveEntry,457,48666,99.069682
114,AIModelsWantEntry,474,48649,99.035075
161,AIAgentOrchWrite,477,48646,99.028968
54,JobSatPoints_15_TEXT,664,48459,98.648291
159,AIAgentKnowWrite,765,48358,98.442685
113,AIModelsHaveEntry,775,48348,98.422328
130,SO_Actions_15_TEXT,823,48300,98.324614
165,AIAgentExtWrite,858,48265,98.253364


In [13]:
print('Аналітичний висновок за первинним дослідженням даних:')
print('- Кількість рядків в основному датасеті:', df.shape[0])
print('- Кількість колонок в основному датасеті:', df.shape[1])
print('- Кількість рядків в схемі питань:', df1.shape[0])
print('- Кількість колонок в схемі питань:', df1.shape[1])
print('- Кількість колонок там, де наявний хоча б один пропуск:', (df.isna().sum() > 0).sum())
print('- Відсоток колонок з пропусками, %:', round((df.isna().sum() > 0).mean() * 100, 2))

Аналітичний висновок за первинним дослідженням даних:
- Кількість рядків в основному датасеті: 49123
- Кількість колонок в основному датасеті: 170
- Кількість рядків в схемі питань: 139
- Кількість колонок в схемі питань: 6
- Кількість колонок там, де наявний хоча б один пропуск: 167
- Відсоток колонок з пропусками, %: 98.24


**Висновок:**  
Stack Overflow - це професійна платформа розробників.  
Представлений датасет - доволі велике міжнародне дослідження.  
Проте він є неоднорідним, тому перед аналізом визначених метрик критично необхідно перевіряти наявні пропуски.

# 1. Підрахунок загальної кількості респондентів.

**Мета завдання:**
визначити, скільки всього респондентів прийняли участь в опитуванні Stack Overflow

In [14]:
# В представленному основному датасеті один рядок має відповідати одному респонденту;
# Кількість рядків можна визначити за допомогою len(df) або df.shape[0];
# Використаємо df.shape[0], т.я. вона працює саме з рядками DataFrame

total_respondents = df.shape[0]

print('Загальна кількість респондентів в опитуванні:', total_respondents)

Загальна кількість респондентів в опитуванні: 49123


In [15]:
# Перевіримо, що в датасеті немає дублікатів респондентів, та один рядок дійсно відповідає одній людині

# Перевірка через унікальні ID
unique_ids = df['ResponseId'].nunique()

print("Кількість рядків / респондентів:", total_respondents)
print("Унікальних ResponseId:", unique_ids)

# Перевірка співпадіння
print("Результати співпадають:", total_respondents == unique_ids)

Кількість рядків / респондентів: 49123
Унікальних ResponseId: 49123
Результати співпадають: True


In [16]:
print('Аналітичний висновок до завдання №1:')
print(f'- В опитуванні представлено {total_respondents} респонденти.')
print('- Це базова метрика обʼєму вибірки.')
print('- Чим більша вибірка, тим стійкіші загальні висновки серед опитуваних.')
print('- В подальших розрахунках цей показник буде використовуватись як знаменник для відсоткових метрик.')

Аналітичний висновок до завдання №1:
- В опитуванні представлено 49123 респонденти.
- Це базова метрика обʼєму вибірки.
- Чим більша вибірка, тим стійкіші загальні висновки серед опитуваних.
- В подальших розрахунках цей показник буде використовуватись як знаменник для відсоткових метрик.


**Висновок:**  
Представлений датасет - результат опитування, в якому прийняли участь понад 49тис. спеціалістів з IT-індустрії.  
Спираючись на наявні дані можна аналізувати міжнародний IT-ринок, заробітні плати, технології, освіту та інтрепретувати результати для виявлення глобальних тенденцій.

# 2. Аналіз повноти відповідей респондентів.

**Мета завдання:**
визначити, скільки респондентів відповіли на **всі питання**, які наявні в schema.

In [17]:
# Хоча ми це вже знаємо, ще раз перевіримо наявність стовпця qname у schema

'qname' in df1.columns

True

In [18]:
# Отримаємо список технічних назв питань зі schema['qname'].
# dropna() прибере пропущені значення в самій schema.
# astype(str) гарантує, що всі назви будуть рядками.
df1_questions = set(df1['qname'].dropna().astype(str))

# Отримаємо множину стовпців основного датасету.
dataset_columns = set(df.columns)

# Знаходимо перетин двох множин.
# intersection() залишає тільки ті qname, які реально існують як стовпці в df.
question_columns = sorted(df1_questions.intersection(dataset_columns))

print('Кількість питань в schema[qname]:', len(df1_questions))
print('Кількість стовпців в основному датасеті:', len(dataset_columns))
print('Кількість питань, знайдених і в schema, і в основному датасеті:', len(question_columns))

# Виводимо перші 20 знайдених стовпців для перевірки.
question_columns[:20]

Кількість питань в schema[qname]: 139
Кількість стовпців в основному датасеті: 170
Кількість питань, знайдених і в schema, і в основному датасеті: 124


['AIAcc',
 'AIAgentChange',
 'AIAgentExtWrite',
 'AIAgentExternal',
 'AIAgentKnowWrite',
 'AIAgentKnowledge',
 'AIAgentObsWrite',
 'AIAgentObserveSecure',
 'AIAgentOrchWrite',
 'AIAgentOrchestration',
 'AIAgent_Uses',
 'AIAgents',
 'AIComplex',
 'AIExplain',
 'AIFrustration',
 'AIHuman',
 'AILearnHow',
 'AIModelsChoice',
 'AIModelsHaveEntry',
 'AIModelsWantEntry']

In [19]:
# Створюємо таблицю тільки з тими стовпцями, які є питаннями зі schema та є в основному датасеті.
answers_to_schema_questions = df[question_columns]

# notna() повертає True, якщо значення НЕ відсутнє.
# all(axis=1) перевіряє умову по всім колонкам в кожному рядку.
# Якщо в усіх питаннях рядка є відповіді, результат буде True.
complete_answers_mask = answers_to_schema_questions.notna().all(axis=1)

# sum() рахує кількість True (тобто 1).
complete_respondents = complete_answers_mask.sum()

print('Кількість респондентів, які відповіли на всі знайдені питання:', complete_respondents)
print('Відсоток таких респондентів, %:', round(complete_respondents / total_respondents * 100, 2))

Кількість респондентів, які відповіли на всі знайдені питання: 0
Відсоток таких респондентів, %: 0.0


In [20]:
print('Аналітичний висновок до завдання №2:')
print(f'- Кількість респондентів без пропусків за всіма знайденими питаннями: {complete_respondents}.')
print(f'- Їх відсоток від загальної вибірки: {round(complete_respondents / total_respondents * 100, 2)}%.')
print('- Низьке або нульве, як в нашому випадку, значення цієї метрики є нормальним для великих опитувань: частина питань може бути не обовʼязковою, опціональною, залежати від попередніх відповідей або пропонуватися лише окремим групам респондентів.')
print('- Для подальшого аналізу неможна автоматично видаляти всі рядки з пропусками, т.я. це може призвести до значного скорочення вибірки та спотворювання висновків.')

Аналітичний висновок до завдання №2:
- Кількість респондентів без пропусків за всіма знайденими питаннями: 0.
- Їх відсоток від загальної вибірки: 0.0%.
- Низьке або нульве, як в нашому випадку, значення цієї метрики є нормальним для великих опитувань: частина питань може бути не обовʼязковою, опціональною, залежати від попередніх відповідей або пропонуватися лише окремим групам респондентів.
- Для подальшого аналізу неможна автоматично видаляти всі рядки з пропусками, т.я. це може призвести до значного скорочення вибірки та спотворювання висновків.


# 3. Стастичний аналіз досвіду респондентів.

**Мета завдання:**
визначити міри центральної тенденції для поля WorkExp

Необхідно знайти:

- середнє значення (mean() - повертає середнє значення, проте чутливе до екстремально великих/малих значень);
- медіану (median() - повертає центральне значення і більш стійке до вибросів);
- моду (mode() - повертає значення, що зустрічається найчастіше).

Pandas за замовчуванням ігнорує пропуски в mean(), median() та mode(), проте, ми додатково перевіримо їх кількість.

In [21]:
# Перевіримо наявність стовпця WorkExp.
'WorkExp' in df.columns

True

In [22]:
# Переведемо стовпець WorkExp у числовий формат.
# Для безпечного переходу використаємо errors='coerce' - якщо значення неможливо конвертувати у число, тоді замінить його на NaN.

work_exp = pd.to_numeric(df['WorkExp'], errors='coerce')

work_exp_missing = work_exp.isna().sum()
work_exp_non_missing = work_exp.notna().sum()

print('Непустих значень WorkExp:', work_exp_non_missing)
print('Пропущених значень WorkExp:', work_exp_missing)
print('Відсоток пропусків WorkExp, %:', round(work_exp_missing / total_respondents * 100, 2))

Непустих значень WorkExp: 42844
Пропущених значень WorkExp: 6279
Відсоток пропусків WorkExp, %: 12.78


In [23]:
# Розрахуємо міри центральної тенденції.

work_exp = df['WorkExp'].dropna() # конкретно в цьому випадку, навіть якщо закоментувати цей пункт - результат не зміниться 

work_exp_mean = round(work_exp.mean(),2)
work_exp_median = work_exp.median()
work_exp_mode = work_exp.mode()

# mode() може повернути декілька значень, якщо ці значення зустрічаються однаково часто.
# щоб перевірити це, оформимо усі знайдені моди у список для виводу у таблиці.
work_exp_mode_list = work_exp_mode.tolist()

work_exp_stats = pd.DataFrame({
    'metric': ['mean', 'median', 'mode'],
    'value': [work_exp_mean, work_exp_median, work_exp_mode_list]
})

work_exp_stats

,metric,value
0,mean,13.37
1,median,10.0
2,mode,[10.0]


In [24]:
print('Аналітичний висновок до завдання №3:')
print(f'- Середній досвід роботи: {round(work_exp_mean, 2)} років.')
print(f'- Медіанний досвід роботи: {round(work_exp_median, 2)} років.')
print(f'- Значення досвіда роботи, що зустрічається найчастіше: {work_exp_mode_list}.')
print('- Для характеристики типового респондента медіанне значення надійніше за середнє, т.я. воно в меншій мірі залежить від викидів.')

Аналітичний висновок до завдання №3:
- Середній досвід роботи: 13.37 років.
- Медіанний досвід роботи: 10.0 років.
- Значення досвіда роботи, що зустрічається найчастіше: [10.0].
- Для характеристики типового респондента медіанне значення надійніше за середнє, т.я. воно в меншій мірі залежить від викидів.


**Висновок:**  
В результаті маємо **значення середнього вище за медіану (13.37>10)**, це свідчить про те, що в датасеті є група спеціалістів з дуже великим періодом 'WorkExp', які "піднімають" середнє значення досвіду по датасету.  
**Медіана та мода збігаються (10=[10])**, тобто типовий спеціаліст у вибірці має досвід роботи близько 10 років, звідси можна зробити висновок, що опитування проходили досвідчені спеціалісти.

# 4. Аналіз віддаленої роботи.

**Мета завдання:**
визначити кількість респондентів, які працюють віддалено

In [25]:
# Перевіримо наявність стовпця RemoteWork.
'RemoteWork' in df.columns

True

In [26]:
# Дивимось розподіл значень в стовпці RemoteWork.
# value_counts(dropna=False) повертає і реальні категорії, і пропуски.
remote_work_distribution = df['RemoteWork'].value_counts(dropna=False)
remote_work_distribution

RemoteWork
NaN                                                                             15376
Remote                                                                          10924
Hybrid (some remote, leans heavy to in-person)                                   6728
In-person                                                                        6034
Hybrid (some in-person, leans heavy to flexibility)                              5824
Your choice (very flexible, you can come in when you want or just as needed)     4237
Name: count, dtype: int64

In [27]:
# Розрахунок строго fully remote:

fully_remote_workers = df[df['RemoteWork'] == 'Remote']

fully_remote_count = fully_remote_workers.shape[0]
fully_remote_percent = round(fully_remote_count / total_respondents * 100, 2)
print('Кількість респондентів, які працюють віддалено:', fully_remote_count)
print('Відсоток від всіх респондентів, %:', fully_remote_percent)

Кількість респондентів, які працюють віддалено: 10924
Відсоток від всіх респондентів, %: 22.24


In [28]:
# Це додатковий розрахунок, через широку фільтрацію він включає часткове співпадіння за критерієм remote.
# Створюємо булеву маску для респондентів, які працюють віддалено.
# astype(str) переводить значення в рядки.
# str.contains() шукає слово remote всередині значення.
# case=False робить пошут нечутливим до регістру.
# na=False захищає код від пропусків.
remote_mask = df['RemoteWork'].astype(str).str.contains('remote', case=False, na=False)

remote_count = remote_mask.sum()
remote_percent = remote_count / total_respondents * 100
dif = round(remote_percent - fully_remote_percent, 2)

print('Кількість респондентів, які працюють віддалено та гібрідно:', remote_count)
print('Відсоток від всіх респондентів, %:', round(remote_percent, 2))
print('Різниця hybrid-remote, %:', dif)

Кількість респондентів, які працюють віддалено та гібрідно: 17652
Відсоток від всіх респондентів, %: 35.93
Різниця hybrid-remote, %: 13.69


In [29]:
print('Аналітичний висновок до завдання №4:')
print(f'- Віддалено працюють {fully_remote_count} респондентів, що складає {fully_remote_percent}% вибірки.')
print('- Ця метрика допомагає оцінити розповсюдженність remote-формата серед розробників.')

Аналітичний висновок до завдання №4:
- Віддалено працюють 10924 респондентів, що складає 22.24% вибірки.
- Ця метрика допомагає оцінити розповсюдженність remote-формата серед розробників.


**Висновок:**  
Датасет демонструє наявність декількох моделей роботи: fully remote, hybrid та flexisble office.  
Remote-формат роботи став повноцінною частиною IT-індустрії.  
Згідно даних датасету в форматі fully-remote працює більше 20% опитуваних, ще 13.69% працюють гібридно (частково віддалено), проте необхідно зазначити, що багато респондентів не дали відповідь на це запитання.  
Для ринку праці високий відсоток віддаленої роботи може означати велику конкуренцію між країнами і більш широкий вибір вакансій для спеціалістів.

# 5. Визначення популярності Python.

**Мета завдання:**
визначити відсоток респондентів, які програмують на Python

In [30]:
# Перевіряємо наявність колонки з вказанням мов програмування.
'LanguageHaveWorkedWith' in df.columns

True

In [31]:
# Створюємо маску респондентів, у яких серед вказаних мов програмування є Python.
# використаємо na=False, пропуски не мають рахуватися як Python та "ламати" пошук.
python_mask = df['LanguageHaveWorkedWith'].str.contains('Python', case=False, na=False)

python_count = python_mask.sum()
python_percent = python_count / total_respondents * 100

print('Кількість респондентів, які використовують Python:', python_count)
print('Відсоток респондентів, які використовують Python:', round(python_percent, 2), '%')

Кількість респондентів, які використовують Python: 18444
Відсоток респондентів, які використовують Python: 37.55 %


In [32]:
print('Аналітичний висновок до завдання №5:')
print(f'- Python використовують {python_count} респондентів, що складає {round(python_percent, 2)}% всієї вибірки.')
print('- Python залишається важливою мовою в IT-індустрії.')
print('- Результати цієї метрики підтверджують практичну цінність вивчення Python.')

Аналітичний висновок до завдання №5:
- Python використовують 18444 респондентів, що складає 37.55% всієї вибірки.
- Python залишається важливою мовою в IT-індустрії.
- Результати цієї метрики підтверджують практичну цінність вивчення Python.


**Висновок:**  
Більше третини опитуваних спеціалістів використовують Python.  
Це досить високий відсоток, що свідчить про те, що ця мова є одним із найпоширеніших і найзатребуваніших мов програмування, яка використовується у великій кількості IT-сфер.

# 6. Аналіз шляхів навчання програмуванню.

**Мета завдання:**
визначити, скільки респондентів вчились програмувати за допомогою онлайн-курсів

In [33]:
# Перевіряємо наявність стовпця LearnCode.
'LearnCode' in df.columns

True

In [34]:
# Ознайомимося з першими непустими значеннями LearnCode, для визначення категорій.
df['LearnCode'].dropna().head(10)

0                                                                                                                                                                                                                                                                                          Online Courses or Certification (includes all media types);Other online resources (e.g. standard search, forum, online community)
1                                                                                                                                                             Online Courses or Certification (includes all media types);Other online resources (e.g. standard search, forum, online community);Books / Physical media;Videos (not associated with specific online course or certification);Stack Overflow or Stack Exchange
2                                                                                                                                                                             

In [35]:
# Шукаємо респондентів, які обрали "Online Courses" як спосіб навчання.

online_courses = df[df['LearnCode'].str.contains('Online Courses', na=False)]

online_courses_count = online_courses.shape[0]

print('Кількість респондентів, які навчались через онлайн-курси:', online_courses_count)

Кількість респондентів, які навчались через онлайн-курси: 10957


In [36]:
print('Аналітичний висновок до завдання №6:')
print(f'- За допомогою онлайн-курсів навчалось {online_courses_count} респондентів')
print('- Для аналізу обрані тільки респонденти, що вказали категорію "Online Courses", інші онлайн-джерела освіти (відео, форуми, документація тощо) не включалися у розрахунок.')
print('- Т.я. респонденти могли обирати декілька варіантів освіти одночасно, використовувалась перевірка наявності категорії "Online Courses" всередині multi-select відповіді.')

Аналітичний висновок до завдання №6:
- За допомогою онлайн-курсів навчалось 10957 респондентів
- Для аналізу обрані тільки респонденти, що вказали категорію "Online Courses", інші онлайн-джерела освіти (відео, форуми, документація тощо) не включалися у розрахунок.
- Т.я. респонденти могли обирати декілька варіантів освіти одночасно, використовувалась перевірка наявності категорії "Online Courses" всередині multi-select відповіді.


**Висновок:**  
Традиційна освіта перестала бути єдиним шляхом вступу в спеціальність.  
Онлайн-освіта є важливим каналом входа в професію та розвитку навичок в IT.  
Онлайн-курси вже є достатньо розповсюдженою та затребуваною формою освіти.

# 7. Географічний аналіз компенсації Python-розробників.

**Мета завдання:**
серед респондентів, які використовують Python, розрахувати середню і медіанну річну компенсацію за країнами

In [37]:
# Перевіримо наявність необхідних стовпців.
# Для корректного аналізу використовуємо вже конвертований в єдину валюту стовпець 'ConvertedCompYearly'.
required_columns_task_7 = ['LanguageHaveWorkedWith', 'Country', 'ConvertedCompYearly']
[col in df.columns for col in required_columns_task_7]

[True, True, True]

In [38]:
# Створимо окрему таблицю тільки з Python-розробниками.
python_developers = df[python_mask].copy()

# Переведемо компенсацію у числовий формат на випадок, якщо наявні текстові значення.
python_developers['ConvertedCompYearly'] = pd.to_numeric(python_developers['ConvertedCompYearly'], errors='coerce')

# Приберемо рядки без країни та без компенсації.
python_compensation_clean = python_developers.dropna(subset=['Country', 'ConvertedCompYearly'])

print('Python-розробників всього:', python_developers.shape[0])
print('Python-розробників із вказаними країною та компенсацією:', python_compensation_clean.shape[0])

Python-розробників всього: 18444
Python-розробників із вказаними країною та компенсацією: 12760


In [39]:
# Згрупуємо дані за країною.
# agg() дозволяє порахувати за один раз декілька агрегованих метрик.
python_comp_by_country = (
    python_compensation_clean
    .groupby('Country')['ConvertedCompYearly']
    .agg(
        respondents_count='count',
        mean_compensation='mean',
        median_compensation='median'
    )
    .reset_index()
)

# Відсортуємо по медіанній компенсації, т.я. вона більш стійка до вибросів.
python_comp_by_country = python_comp_by_country.sort_values('median_compensation', ascending=False)

python_comp_by_country.head(20)

,Country,respondents_count,mean_compensation,median_compensation
102,Oman,2,390135.000000,390135.0
3,Andorra,2,226103.500000,226103.5
145,United States of America,3124,173295.260243,150000.0
65,Israel,104,135828.365385,142594.0
132,Switzerland,240,156456.600000,142592.0
99,Nomadic,7,120131.571429,139218.0
63,Ireland,74,120523.918919,116015.0
78,Luxembourg,7,116014.714286,109054.0
73,Kyrgyzstan,2,106008.500000,106008.5
15,Belize,1,102121.000000,102121.0


**Висновок:**  
**Медіанна компенсація** Python-розробників доволі висока, і в **USA складає 150000** (умовних одиниць).  
Високі показники медіанної компенсації для країн **Oman та Andorra** необхідно аналізувати з обережністю, оскільки вибірка **(count = 2)** дуже маленька та недостатньо показова.  
Географічний аналіз компенсації допомагає порівнювати ринки праці, проте слід враховувати вартість життя, податків та інші критерії.

# 8. Аналіз освіти найбільш оплачуваних спеціалістів.

**Мета завдання:**
визначити рівні освіти 5 респондентів з найбільшою компенсацією

In [40]:
# Перевіряємо наявність необхідних стовпців.
required_columns_task_8 = ['ConvertedCompYearly', 'EdLevel']
[col in df.columns for col in required_columns_task_8]

[True, True]

In [41]:
# Створимо копію з необхідними колонками.
# Додамо стовпець Country (якщо є в датасеті), для кращого аналізу.
columns_for_top_comp = ['ConvertedCompYearly', 'EdLevel']
if 'Country' in df.columns:
    columns_for_top_comp.append('Country')

compensation_education = df[columns_for_top_comp].copy()
compensation_education['ConvertedCompYearly'] = pd.to_numeric(compensation_education['ConvertedCompYearly'], errors='coerce')

# Приберемо рядки без компенсації, т.я. їх не можна коректно сортувати.
compensation_education_clean = compensation_education.dropna(subset=['ConvertedCompYearly'])

# Сортуємо за спаданням та обираємо ТОП-5.
top_5_paid = compensation_education_clean.sort_values('ConvertedCompYearly', ascending=False).head(5)

top_5_paid

,ConvertedCompYearly,EdLevel,Country
34267,50000000.0,"Associate degree (A.A., A.S., etc.)",United States of America
28700,33552715.0,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Ukraine
43143,18387548.0,"Associate degree (A.A., A.S., etc.)",Poland
35353,15430267.0,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)","Iran, Islamic Republic of..."
45971,13921760.0,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Netherlands


In [42]:
print('Аналітичний висновок до завдання №8:')
print('- Таблиця демонструє 5 респондентів з максимальною вказаною річною компенсацією та їх рівень освіти.')
print('- За цим ТОП-5 не можна робити загальний висновок про вплив освіти на дохід, т.я. це достатньо маленька вибірка.')
print('- Цей результат показує скоріш індивідуальні екстремальні значення, тому висновки необхідно робити обережно.')
print('- Для більш надійного аналізу освіти та компенсації необхідно згрупувати всіх респондентів по EdLevel та порівняти медіанні значення.')

Аналітичний висновок до завдання №8:
- Таблиця демонструє 5 респондентів з максимальною вказаною річною компенсацією та їх рівень освіти.
- За цим ТОП-5 не можна робити загальний висновок про вплив освіти на дохід, т.я. це достатньо маленька вибірка.
- Цей результат показує скоріш індивідуальні екстремальні значення, тому висновки необхідно робити обережно.
- Для більш надійного аналізу освіти та компенсації необхідно згрупувати всіх респондентів по EdLevel та порівняти медіанні значення.


**Висновок:**  
**Associate degree (A.A., A.S., etc.)** - Associate of Arts, Associate of Science -  
це дворічна середня професійна освіта в США;  
**Bachelor’s degree (B.A., B.S., B.Eng., etc.)** - Bachelor of Arts, Bachelor of Science, Bachelor of Engineering -  
це повноцінна вища освіта рівня бакалавра;  
**Master’s degree (M.A., M.S., M.Eng., MBA, etc.)** - Master of Arts, Master of Science, Master of Engineering, Master of Business Administration -  
це магістерський рівень освіти.  
Серед спеціалістів із максимальною річною компенсацією наявні **НЕ лише магістри**. Тобто висока компенсація НЕ гарантується виключно "академічною освітою", проте магістратура зустрічається доволі часто.  

Щодо самих значень річної компенсації, представлених в ТОП-5, можливо, але не точно, це можуть бути виброси: або некорректне введення даних, або дані від власників бізнесу. В будь-якому випадку я б не посилался на ці значення без додаткового дослідження.

**8.1. Додаткова перевірка до Завдання №8 -  
медіанна компенсація за рівнем освіти.**

In [43]:
# Згрупуємо всіх респондентів із вказаною компенсацією за рівнем освіти.
# Розрохуємо кількість, середнє та медіану.
education_comp_summary = (
    compensation_education_clean
    .dropna(subset=['EdLevel'])
    .groupby('EdLevel')['ConvertedCompYearly']
    .agg(
        respondents_count='count',
        mean_compensation='mean',
        median_compensation='median'
    )
    .reset_index()
    .sort_values('median_compensation', ascending=False)
)

education_comp_summary

,EdLevel,respondents_count,mean_compensation,median_compensation
5,"Professional degree (JD, MD, Ph.D, Ed.D, etc.)",1362,128251.733480,87011.0
2,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",6913,105089.565890,80000.0
1,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",10417,98194.930690,75744.0
0,"Associate degree (A.A., A.S., etc.)",797,169131.329987,69751.0
7,Some college/university study without earning a degree,2860,87739.438112,66033.5
6,"Secondary school (e.g. American high school, German Realschule or Gymnasium, etc.)",1213,71322.877988,52207.0
3,Other (please specify):,222,89789.792793,49943.0
4,Primary/elementary school,127,137027.826772,46646.0


In [44]:
print('Додатковий аналітичний висновок за освітою та компенсацією:')
print('- Групування за EdLevel надає більш надійні дані, ніж ТОП-5 окремих респондентів.')
print('- Для порівняння рівнів освіти краще розглядати медіану, а не тільки середнє.')
print('- Кількість респондентів в кожній групі також необхідно враховувати: маленькі групи можуть давати нестабільні значення.')

Додатковий аналітичний висновок за освітою та компенсацією:
- Групування за EdLevel надає більш надійні дані, ніж ТОП-5 окремих респондентів.
- Для порівняння рівнів освіти краще розглядати медіану, а не тільки середнє.
- Кількість респондентів в кожній групі також необхідно враховувати: маленькі групи можуть давати нестабільні значення.


**Висновок:**  
Тут ще раз наочно видно "розмах" рівнів освіти спеціалістів, проте вже доволі чітко прослідковується залежність медіанної компенсації від рівня освіти (сортування виконувалось саме за медіанною компенсацією).

# 9. Аналіз популярності Python по віковим категоріям.

**Мета завдання:**
для кожної вікової категорії  визначити відсоток респондентів, які програмують на Python

In [45]:
# Перевіряємо наявність стовпця Age.
'Age' in df.columns

True

In [46]:
# Створимо таблицю для аналізу Python за віком.
age_python = pd.DataFrame({
    'Age': df['Age'],
    'uses_python': python_mask
})

# Приберемо рядки, в яких не вказаний вік.
age_python_clean = age_python.dropna(subset=['Age'])

# Групуємо за віком та розраховуємо метрики.
python_by_age = (
    age_python_clean
    .groupby('Age')['uses_python']
    .agg(
        respondents_count='count',
        python_users='sum'
    ).reset_index())

# Т.я. uses_python — булеве значення, sum() рахує кількість True.
python_by_age['python_percent'] = python_by_age['python_users'] / python_by_age['respondents_count'] * 100

# Відсортуємо за відсотком Python-розробників за спаданням.
python_by_age = python_by_age.sort_values('python_percent', ascending=False)

python_by_age

,Age,respondents_count,python_users,python_percent
0,18-24 years old,9195,3681,40.032626
3,45-54 years old,6265,2422,38.659218
4,55-64 years old,2626,978,37.242955
1,25-34 years old,16485,6090,36.942675
2,35-44 years old,13232,4857,36.706469
5,65 years or older,942,298,31.634820
6,Prefer not to say,378,118,31.216931


In [47]:
print('Аналітичний висновок до завдання №9:')
print('- Таблиця демонструє відсоток Python-розробників по кожній віковій категорії.')

Аналітичний висновок до завдання №9:
- Таблиця демонструє відсоток Python-розробників по кожній віковій категорії.


**Висновок:**  
Python приблизно однаково популярний серед усіх вікових категорій, до того ж відсоток Python-розробників достатньо високий, що ще раз підтверджує популярність мови Python серед спеціалістів.

# 10. Аналіз індустрій серед високооплачуваних віддалених працівників.

**Мета завдання:**
знайти респондентів, які одночасно:
1. знаходяться вище 75-го перцентиля компенсації;
2. працюють віддалено;
3. мають вказану індустрію.  
Визначити найрозповсюдженіші індустрії в цій групі

In [48]:
# Перевіримо наявність необхідних колонок.
required_columns_task_10 = ['ConvertedCompYearly', 'RemoteWork', 'Industry']
[col in df.columns for col in required_columns_task_10]

[True, True, True]

In [49]:
# Створимо таблицю для аналізу та переведемо стовпець компенсації у числовий формат.
high_paid_remote = df[['ConvertedCompYearly', 'RemoteWork', 'Industry']].copy()
high_paid_remote['ConvertedCompYearly'] = pd.to_numeric(high_paid_remote['ConvertedCompYearly'], errors='coerce')

# Розрахуємо 75-й перцентиль тільки по непустим значенням компенсації.
# 75-й перцентиль визначає поріг, вище якого знаходиться верхня чверть значень по компенсації.
compensation_75_percentile = high_paid_remote['ConvertedCompYearly'].quantile(0.75)

print('75-й перцентиль річної компенсації:', compensation_75_percentile)

75-й перцентиль річної компенсації: 120630.25


In [50]:
# Створимо декілька умов фільтрації.
# 1. Компенсація вище або дорівнює 75-му перцентилю.
condition_high_comp = high_paid_remote['ConvertedCompYearly'] >= compensation_75_percentile

# 2. Формат роботи - строго fully remote.
condition_remote = (high_paid_remote['RemoteWork'] == 'Remote')

# 3. Індустрія вказана респондентом.
condition_industry_not_missing = high_paid_remote['Industry'].notna()

# Комбінуємо умови за домопогою &.
high_paid_remote_industry = high_paid_remote[condition_high_comp & condition_remote & condition_industry_not_missing]

print('Кількість високооплачуваних віддалених робітників із вказаною індустрією:', high_paid_remote_industry.shape[0])

Кількість високооплачуваних віддалених робітників із вказаною індустрією: 2456


In [51]:
# Рахуємо найпоширеніші індустрії.
# value_counts() сортує значення за частотою від більшого до меншого.
top_industries_high_paid_remote = (
    high_paid_remote_industry['Industry']
    .value_counts()
    .reset_index())

top_industries_high_paid_remote.columns = ['Industry', 'respondents_count']

top_industries_high_paid_remote.head(20)

,Industry,respondents_count
0,Software Development,1184
1,Fintech,189
2,Healthcare,188
3,Other:,176
4,"Internet, Telecomm or Information Services",138
5,Banking/Financial Services,88
6,Government,78
7,Media & Advertising Services,75
8,Retail and Consumer Services,64
9,"Transportation, or Supply Chain",63


In [52]:
print('Аналітичний висновок до завдання №10:')
print('- Визначено респондентів із високою компенсацією при fully remote-форматі роботи та вказаною індустрією.')
print('- Результат демонструє, в яких індустріях найчастіше зустрічаються високооплачувані спеціалісти, які працюють віддалено.')

Аналітичний висновок до завдання №10:
- Визначено респондентів із високою компенсацією при fully remote-форматі роботи та вказаною індустрією.
- Результат демонструє, в яких індустріях найчастіше зустрічаються високооплачувані спеціалісти, які працюють віддалено.


**Висновок:**  
Посилаючись на дані датасету, **Software Development** — абсолютний лідер серед усіх індустрій для високооплачуваних спеціалістів, що працюють віддалено.  
Проте, для фінальних висновків важливо враховувати не тільки кількість, але і структуру всієї вибірки: великі індустрії можуть частіше попадати в топ через більше число респондентів.

**Фінальний висновок:**  
Дослідження показує декілька фундаментальних тенденцій сучасного IT-ринка (станом на 2025 рік):  
IT-сфера - це глобальна індустрія, яка дозволяє працювати віддалено;  
Python - універсальна, затребувана та високоплачувана мова програмування;  
Найбільш високооплачувані спеціалісти мають різні освітні рівні, з чого можна зробити висновок, що практичні навички часто важливіше формальної освіти;  
Онлайн-освіта вже стала повноцінним способом входу в індустрію.  
Реальні дані завжди містять пропуски, викиди та неідеальні відповіді, тому на кожному етапі важливі критична інтерпретація даних та перевірка коректності висновків.